### ISSUE TITLE
v2

### ISSUE THREAD

**Issue by 89z** [jhump/protoreflect#418]
The v2 package is here:

~~~
google.golang.org/protobuf
~~~

and the v1 package is here:

~~~
github.com/golang/protobuf
~~~

I notice that the v1 package is still being used:

https://github.com/jhump/protoreflect/blob/db1a327a393ec12e4a565f4813c154abd843376a/desc/descriptor.go#L12-L13

are any plans to change to v2?

**Comment by jhump** [comment]
Yep -- the biggest piece of which is in #354. But it's rather high-risk at the moment due to the impact to the `protoparse` package, which has a number of major consumers. That's why the PR has been open, unmerged, for nearly a year.

Note that the new v2 API largely _deprecates_ the need to use `protoreflect` -- at least the `desc` and `dynamic` packages. And, in the meantime, this repo is still _compatible_ with V2, so you can use this repo from code that uses v2.

I'm currently working on a separate project to overhaul protoparse and extract it into a separate repo. Then it should be much safer to land the PR that moves this repo to the v2 APIs. In fact, after that, I will probably introduce a v2 of this repo's APIs and try to deprecate much of what's in here now (since it is redundant with capabilities in the protobuf v2 API).

**Comment by jhump** [comment]
> If thats the case, it really seems strange to me, as you can certainly do it with JSON [3], XML [4] and CSV [5].

What does this even mean? You can parse `.proto` files into JSON or XML? Or are you suggesting that `.proto` is just a data encoding, like JSON or XML? If the latter, I'm afraid it is _not_ that. It is an IDL -- it's a source program that describes data, not a format for encoding data.

> I know about Protoc [6], but that just seems like such a backward way to do what I want to do. 

`protoc` is the official way to compile protos. I hardly think this is backwards -- why does it matter what language it was written in? All of the unix command-line tools are written in C, but I see nothing wrong with using them even though I don't personally care to write C code.

> So I just wanted to make sure that in fact, the main Go package cannot do what I want to do

I don't actually understand what you are trying to do.

It is true that the official Go runtime for protobufs _cannot_ parse/compile `.proto` source files. So if you _need_ to do that, then the protoparse package in this repo is what you want.

However, I don't quite understand why you need to parse/compile `.proto` files. The much more typical way to work with protos is to generate code from them -- that's where you'd use `protoc`, to _generate_ Go code that lets you work with the data and services described in your `.proto` source file.

**Comment by 89z** [comment]
> However, I don't quite understand why you need to parse/compile `.proto` files. The much more typical way to work with protos is to generate code from them -- that's where you'd use `protoc`, to _generate_ Go code that lets you work with the data and services described in your `.proto` source file.

So I was not clear. Let me try again. With JSON for example, you can use a flow like this:

~~~go
package date
import "encoding/json"

type date struct {
   Month int
   Day int
}

func newDate(b []byte) (*date, error) {
   d := new(date)
   if err := json.Unmarshal(b, d); err != nil {
      return nil, err
   }
   return d, nil
}
~~~

So you define the type, then you Unmarshal wire format data into the type. I want to do the same thing with Protobuf. So ideally, I would make a `date.proto`:

~~~proto
message Date {
   optional int64 month = 1;
   optional int64 day = 2;
}
~~~

Then Unmarshal that into a Message, then Unmarshal some wire format data into the message. I think what I am wanting to do is possible, but I dont want the Protoc tool to be used. Given that, it seems I cannot use the `google.golang.org/protobuf` to make that happen, but I could use `protoparse`. If I am wrong about above details let me know, thanks again.

**Comment by jhump** [comment]
Using `protoparse` will be extremely inefficient. I still don't understand a need for `protoparse` here: I think you would be _much better_ served by using the normal flow with protobufs. Generate Go code from your `.proto` source (using `protoc`) and then use those generated Go structs to marshal/unmarshal (with the normal Go protobuf runtime library).

The reason you need the compilation step with `protoc` for protos but not for arbitrary data structures with JSON is because the proto format relies on a strongly-typed message schema (and takes advantage of that in the marshaling/unmarshaling so that it uses less CPU to encode/decode and is more compact).

By using `protoparse`, you will repeatedly pay the cost to compile the `.proto` sources, at runtime, and must use a dynamic message (one whose schema is not known at compile-time) in order to marshal/unmarshal, which is less efficient than using the code generated by `protoc`.

Having said all that, there is another alternative to using `protoc` -- a toolchain that offers a lot above and beyond `protoc` with more to come (and it happens to be implemented in Go, not C++): [Buf](https://buf.build/). Go take a look!

Note that Buf still involves a compilation phase to generate Go code (instead of pushing the proto parsing/compilation step to runtime).

**Comment by 89z** [comment]
> Using `protoparse` will be extremely inefficient. I still don't understand a need for `protoparse` here: I think you would be _much better_ served by using the normal flow with protobufs. Generate Go code from your `.proto` source (using `protoc`) and then use those generated Go structs to marshal/unmarshal (with the normal Go protobuf runtime library).

I understand the normal process is the recommended one, but I dont like it, and I dont want to use it. I want to Unmarshal `.proto` files using Go code, not some C++ code, and not a C++ tool. I understand that this comes with a performance hit, but I dont care. I think my usecase is probably simple enough thats a little extra computation is not going to matter. Regardless, I would like to make that decision for myself. You could very well be correct, and doing it "my way" could be awful, even for a simple program. But I want to be the one to make that decision. I want to do tests "my way" versus "the normal way", and if theyre pretty close performance wise, then I do "my way". But if "my way" sucks, then I will use Protoc.

But what frustrates me, is that ever webpage I go to, and every person I talk to says "no, only use Protoc, its the only way". Well, its not the only way. It might be the best way, but people should have option to choose "slow method", if it works for them.

> By using `protoparse`, you will repeatedly pay the cost to compile the `.proto` sources, at runtime, and must use a dynamic message (one whose schema is not known at compile-time) in order to marshal/unmarshal, which is less efficient than using the code generated by `protoc`.

Yes I understand this, see above.

> Having said all that, there is another alternative to using `protoc` -- a toolchain that offers a lot above and beyond `protoc` with more to come (and it happens to be implemented in Go, not C++): [Buf](https://buf.build/). Go take a look!

Yes, I did know about this already. Its appealing, as its a Go tool, but its still a tool. I dont prefer to precompile code like this if possible, as generated code is usually ugly and not as efficient as handwritten code. Granted, its still probably much faster than what Im trying to do, but thats my preference.

**Comment by jhump** [comment]
>  Unmarshal .proto files using Go code, not some C++ code

Just to be clear, Protobuf is an IDL. You aren't unmarshaling it, you are _compiling it_, just like you'd use the `go` tool to compile Go source code (or perhaps more aptly, like you'd use `tsc` to transpile TypeScript to JavaScript). The output of the compiler is something that can then be used to marshal/unmarshal data that uses the schema defined by the `.proto` sources.

> But what frustrates me, is that every webpage I go to, and every person I talk to says "no, only use Protoc, its the only way".

That's because before `protoparse`, that _was_ the only way. And even now, `protoc` _is the authoritative definition_ of the protobuf language. There is not a formal language spec that can be used to build alternatives. That's why you just don't see other implementations. I've put a lot of work into `protoparse` to make its behavior match `protoc` as closely as possible, but without a real spec much of this was reverse engineering and continuous refinement. There likely are other possible options (though maybe mostly in languages other than Go), but they have varying degrees of quality and conformance to "the spec" (e.g. mimic'ing behavior of `protoc`)... I like to think that `protoparse` is the closest to `protoc`.

**Comment by 89z** [comment]
@jhump OK, thanks again. `protoparse` is appealing, but I dont like that its using the old API. However after doing more searching, not many options are even available to do what I am wanting, so maybe its a beggars/choosers situation. I will have to think about what I want to do. Also I did find this:

https://github.com/jhump/goprotoc

Its also a tool, but Im wondering why you recommended Buf over your own tool? Is yours still alpha or something?

**Comment by jhump** [comment]
> Its also a tool, but Im wondering why you recommended Buf over your own tool? Is yours still alpha or something?

Something like that... That repo is mostly valuable for writing `protoc` _plugins_ in Go. There is an attempt at a `protoc` replacement in pure Go, but that part of that repo is definitely pre-alpha. Under-the-hood `goprotoc` uses protoparse.

FWIW, Buf is also powered by protoparse. I try to work pretty closely with them on improvements/fixes to protoparse since they have a very large corpus of proto source files for which their tool is used, and thus they (and their users) are best positioned to find discrepancies between protoparse and `protoc` and report them back to me :)

**Comment by 89z** [comment]
After much looking, I think I found a package that does what I want:

~~~
package main

import (
   "fmt"
   "github.com/philpearl/plenc"
)

type appDetails struct {
   Version string `plenc:"4"`
}

func main() {
   b := []byte("\"\vhello world")
   var m appDetails
   if err := plenc.Unmarshal(b, &m); err != nil {
      panic(err)
   }
   fmt.Printf("%+v\n", m) // {Version:hello world}
}
~~~

I appreciate all the detailed responses, I just think the `protoc` route is not right for me. All I really want, is something to unmarshal Protobuf bytes into a `struct`, similar to `encoding/json` or `encoding/xml`. Its pretty frustrating that the standard packages cant do that, and its hard to find anyone that even interesting in doing something like that. I know Protobuf is not JSON, but most of the cases where Ive seen it used, servers are basically treating Protobuf as a compressed JSON response. So it was especially strange that given the similarity of server usage, that package implementations versus JSON would be so different. Thanks again.
